broker-core/src/main/java/io/zeebe/broker/workflow/state/ElementInstanceState.java

```
@@ -327,6 +327,29 @@ public void updateFailedToken(IndexedRecord indexedRecord) {
     return records;
   }
 
+  @FunctionalInterface
+  public interface TokenVisitor {
+    void visitToken(IndexedRecord indexedRecord);
+  }
+
+  public void visitFailedTokens(long scopeKey, TokenVisitor visitor) {
+    visitTokens(scopeKey, Purpose.FAILED_TOKEN, visitor);
+  }
+
+  private void visitTokens(long scopeKey, Purpose purpose, TokenVisitor visitor) {
+    longKeyPurposeBuffer.putLong(0, scopeKey, STATE_BYTE_ORDER);
+    longKeyPurposeBuffer.putByte(Long.BYTES, (byte) purpose.ordinal());
+
+    rocksDbWrapper.whileEqualPrefix(
+        tokenParentChildHandle,
+        longKeyPurposeBuffer.byteArray(),
+        (key, value) -> {
+          final StoredRecord tokenEvent =
+              getTokenEvent(getLong(key, Long.BYTES + BitUtil.SIZE_OF_BYTE));
```

Add a `null` check to fail fast if the state is corrupted.